### PR TITLE
feat(#1500): Add artifacts-unzip-worker

### PR DESCRIPTION
**Pull request by sakka2**
## Context

To provide the SD_ZIP_ARTIFACTS function by only using Screwdriver components ( According to the [SD ZIP ARTIFACTS design doc](https://github.com/screwdriver-cd/screwdriver/blob/0962d20d251c0b7c03d40b7e4b10bb1ed009831d/design/sd-zip-artifacts.md) ).

## Objective

This PR adds a script & unit test for SD_ZIP_ARTIFACTS function as follows;
- Dequeue Jobs from the Redis Server using node-resque.
- Download a zip artifact from the store.
- Decompress the zip artifact.
- Upload all the decompressed artifacts to the store.
- Delete the downloaded zip artifact.（TBD)

## References

[Issue:#1500](https://github.com/screwdriver-cd/screwdriver/issues/1500)
[Design](https://github.com/screwdriver-cd/screwdriver/blob/0962d20d251c0b7c03d40b7e4b10bb1ed009831d/design/sd-zip-artifacts.md)

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

## TODO -> SOLVED
Retry Plugin in node-resque does not work now.
Even if the job fails, it dose not retry.
reEnqueue seems to be succeeded, but reEnqueued job dose not start.

<details>
<summary>worker's log</summary>

```
worker_1  | {"level":"info","message":"worker[1] check in @ 1638868136","timestamp":"2021-12-07T09:08:55.687Z"}
worker_1  | {"level":"info","message":"worker[1] polling unzip","timestamp":"2021-12-07T09:08:55.690Z"}
worker_1  | {"level":"info","message":"worker[1] working job unzip {\"class\":\"unzip\",\"queue\":\"unzip\",\"args\":[{\"buildId\":13,\"token\":\"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6MTMsInNjb3BlIjpbImJ1aWxkIl0sImlhdCI6MTYzODg2ODEzMywiZXhwIjoxNjM4ODY4NDMzLCJqdGkiOiIxZDcxNWI1Yy0wZDEyLTQ1MGMtYTVmZC01ZmRjNDM4YjQ3NjMifQ.BbKKe90hRoj0u2y-C95zc8_71cvckt2CcD-EXGBdinIMrKcw3mq8cXlvV4oqEytbCEt68hV9uwztoDuHClnUteWCuUfyp3BEreEkGqW63rsTPgxAaOsdoQDXSsYVSxP7BH0gIpdn8kqO4KhU2BSoLBNOga7jBNdSGCSP1NWZ394\"}]}","timestamp":"2021-12-07T09:08:55.692Z"}
worker_1  | {"level":"info","message":"Job:unzip started. buildId: 13","timestamp":"2021-12-07T09:08:55.693Z"}
worker_1  | {"level":"info","message":"worker[2] started","timestamp":"2021-12-07T09:08:55.711Z"}
worker_1  | {"level":"error","message":"Response code 401 (Unauthorized)","timestamp":"2021-12-07T09:08:55.720Z"}
worker_1  | {"level":"info","message":"worker[2] check in @ 1638868136","timestamp":"2021-12-07T09:08:55.722Z"}
worker_1  | {"level":"info","message":"worker[2] polling unzip","timestamp":"2021-12-07T09:08:55.724Z"}
worker_1  | {"level":"info","message":"*** checked for worker status: + (event loop delay: 2ms)","timestamp":"2021-12-07T09:08:55.725Z"}
worker_1  | {"level":"info","message":"worker[2] paused","timestamp":"2021-12-07T09:08:55.726Z"}
worker_1  | {"level":"info","message":"worker[1] reEnqueue job ([object Object]) unzip {\"plugins\":[null],\"pluginOptions\":{\"Retry\":{\"retryLimit\":3,\"retryDelay\":5000,\"backoffStrategy\":null}},\"args\":[{\"buildId\":13,\"token\":\"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6MTMsInNjb3BlIjpbImJ1aWxkIl0sImlhdCI6MTYzODg2ODEzMywiZXhwIjoxNjM4ODY4NDMzLCJqdGkiOiIxZDcxNWI1Yy0wZDEyLTQ1MGMtYTVmZC01ZmRjNDM4YjQ3NjMifQ.BbKKe90hRoj0u2y-C95zc8_71cvckt2CcD-EXGBdinIMrKcw3mq8cXlvV4oqEytbCEt68hV9uwztoDuHClnUteWCuUfyp3BEreEkGqW63rsTPgxAaOsdoQDXSsYVSxP7BH0gIpdn8kqO4KhU2BSoLBNOga7jBNdSGCSP1NWZ394\"}]}","timestamp":"2021-12-07T09:08:55.730Z"}
worker_1  | {"level":"info","message":"worker[1] job success unzip {\"class\":\"unzip\",\"queue\":\"unzip\",\"args\":[{\"buildId\":13,\"token\":\"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6MTMsInNjb3BlIjpbImJ1aWxkIl0sImlhdCI6MTYzODg2ODEzMywiZXhwIjoxNjM4ODY4NDMzLCJqdGkiOiIxZDcxNWI1Yy0wZDEyLTQ1MGMtYTVmZC01ZmRjNDM4YjQ3NjMifQ.BbKKe90hRoj0u2y-C95zc8_71cvckt2CcD-EXGBdinIMrKcw3mq8cXlvV4oqEytbCEt68hV9uwztoDuHClnUteWCuUfyp3BEreEkGqW63rsTPgxAaOsdoQDXSsYVSxP7BH0gIpdn8kqO4KhU2BSoLBNOga7jBNdSGCSP1NWZ394\"}]} >> null","timestamp":"2021-12-07T09:08:55.733Z"}
```
</details>

<details>
<summary>redis keys status after job fails</summary>

```
root@8e26707c7c68:/data# redis-cli
127.0.0.1:6379> keys *
 1) "resque:delayed:1638868141"
 2) "resque:worker:538f4a3e1baf:19+1:unzip:started"
 3) "resque:failure-resque-retry:unzip:{\"buildId\":13,\"token\":\"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6MTMsInNjb3BlIjpbImJ1aWxkIl0sImlhdCI6MTYzODg2ODEzMywiZXhwIjoxNjM4ODY4NDMzLCJqdGkiOiIxZDcxNWI1Yy0wZDEyLTQ1MGMtYTVmZC01ZmRjNDM4YjQ3NjMifQ.BbKKe90hRoj0u2y-C95zc8_71cvckt2CcD-EXGBdinIMrKcw3mq8cXlvV4oqEytbCEt68hV9uwztoDuHClnUteWCuUfyp3BEreEkGqW63rsTPgxAaOsdoQDXSsYVSxP7BH0gIpdn8kqO4KhU2BSoLBNOga7jBNdSGCSP1NWZ394\"}"
 4) "resque:timestamps:{\"class\":\"unzip\",\"queue\":\"unzip\",\"args\":[{\"buildId\":13,\"token\":\"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6MTMsInNjb3BlIjpbImJ1aWxkIl0sImlhdCI6MTYzODg2ODEzMywiZXhwIjoxNjM4ODY4NDMzLCJqdGkiOiIxZDcxNWI1Yy0wZDEyLTQ1MGMtYTVmZC01ZmRjNDM4YjQ3NjMifQ.BbKKe90hRoj0u2y-C95zc8_71cvckt2CcD-EXGBdinIMrKcw3mq8cXlvV4oqEytbCEt68hV9uwztoDuHClnUteWCuUfyp3BEreEkGqW63rsTPgxAaOsdoQDXSsYVSxP7BH0gIpdn8kqO4KhU2BSoLBNOga7jBNdSGCSP1NWZ394\"}]}"
 5) "resque:queues"
 6) "resque:connection_test_key"
 7) "resque:workers"
 8) "resque:worker:7f1c08848e75:20+1:unzip:started"
 9) "resque:stat:processed:7f1c08848e75:20+1"
10) "resque:stat:processed"
11) "resque:delayed_queue_schedule"
12) "resque:stat:processed:538f4a3e1baf:19+1"
13) "resque:stat:failed"
14) "resque:resque-retry:unzip:{\"buildId\":13,\"token\":\"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6MTMsInNjb3BlIjpbImJ1aWxkIl0sImlhdCI6MTYzODg2ODEzMywiZXhwIjoxNjM4ODY4NDMzLCJqdGkiOiIxZDcxNWI1Yy0wZDEyLTQ1MGMtYTVmZC01ZmRjNDM4YjQ3NjMifQ.BbKKe90hRoj0u2y-C95zc8_71cvckt2CcD-EXGBdinIMrKcw3mq8cXlvV4oqEytbCEt68hV9uwztoDuHClnUteWCuUfyp3BEreEkGqW63rsTPgxAaOsdoQDXSsYVSxP7BH0gIpdn8kqO4KhU2BSoLBNOga7jBNdSGCSP1NWZ394\"}"
15) "resque:worker:ping:538f4a3e1baf:19+1"
16) "resque:stat:failed:538f4a3e1baf:19+1"
17) "resque:worker:ping:7f1c08848e75:20+1"
127.0.0.1:6379>
```

</details>

Retry Plugin in my source: https://github.com/sonic-screwdriver-cd/artifacts-unzip-worker/blob/8dde7c7355c4ca2aed46d0759d6f620f513a0bbd/lib/jobs.js#L52

Example of Retry Plugin (node-resque official) : 
https://github.com/actionhero/node-resque/blob/fe89056671af32e35695332b248c1df1e422431e/examples/retry.ts#L27-L45

### CODE REVIEW

File: config/custom-environment-variables.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# All values in this document are the ENVIRONMENT variable names that can override the defaults
+# from `default.yaml`
```
Comment (yuichi10) [inline config/custom-environment-variables.yaml:2]: Do you need to add worker settings?

File: index.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+multiWorker.on('error', (workerId, queue, job, error) => {
+    logger.info(`worker[${workerId}] error ${queue} ${JSON.stringify(job)} >> ${error}`);
+});
+multiWorker.on('pause', workerId => {
+    logger.info(`worker[${workerId}] paused`);
+});
+
+// multiWorker emitters
+multiWorker.on('internalError', error => {
+    logger.info(error);
```
Comment (yuichi10) [inline index.js]:
```suggestion
    logger.error(error);
```

File: lib/jobs.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        await Promise.all(
+            zipEntries.map(async zipEntry => {
+                const fileName = zipEntry.entryName;
+                const file = Buffer.from(zipEntry.getData());
+
+                return store.putArtifact(config.buildId, config.token, fileName, file);
+            })
+        );
+
+        return null;
```
Comment (yuichi10) [inline lib/jobs.js]: Is this necessary?

File: config/custom-environment-variables.yaml (hunk 1 above)
Comment (sakka2) [reply config/custom-environment-variables.yaml:2]: thanks, added worker settings

File: lib/jobs.js (hunk 3 above)
Comment (sakka2) [reply lib/jobs.js]: Maybe no. deleted.

**Comment by sakka2** [comment]
Now, [Retry Plugin](https://github.com/sonic-screwdriver-cd/artifacts-unzip-worker/blob/8dde7c7355c4ca2aed46d0759d6f620f513a0bbd/lib/jobs.js#L52) dose not work.
In the description, I added the unzip-worker's log and redis-server's keys status while the job fails.
Does anyone have any idea with this problem?

File: config/custom-environment-variables.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  redisConnection:
+    host: REDIS_HOST
+    port: REDIS_PORT
+    options:
+      password: REDIS_PASSWORD
+      tls: REDIS_TLS_ENABLED
+    database: REDIS_DB_NUMBER
+  prefix: REDIS_QUEUE_PREFIX
+
+worker:
```
Comment (tk3fftk) [inline config/custom-environment-variables.yaml]: I think the word `worker` could be a bit misunderstandable. there are some workers, unzip-worker or resque worker (or buildcluster-queueworker?)

File: lib/helper/request-store.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ */
+const getZipArtifact = async (buildId, token) => {
+    const options = {
+        url: `${baseUrl}/builds/${buildId}/ARTIFACTS/${ZIP_FILE}`,
+        method: 'GET',
+        headers: {
+            Authorization: token
+        },
+        retry: {
+            limit: 5
```
Comment (tk3fftk) [inline lib/helper/request-store.js]: should we make this to a const?

File: lib/jobs.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
+'use strict';
+
+const logger = require('screwdriver-logger');
+const plugins = require('node-resque').Plugins;
+const AdmZip = require('adm-zip');
+
+const store = require('./helper/request-store');
+
+const RETRY_LIMIT = 3;
+// This is in milliseconds, reference: https://github.com/taskrabbit/node-resque/blob/master/lib/plugins/Retry.js#L12
```
Comment (tk3fftk) [inline lib/jobs.js]: has this moved?
```suggestion
// This is in milliseconds, reference: https://github.com/actionhero/node-resque/blob/fe89056671af32e35695332b248c1df1e422431e/src/plugins/Retry.ts#L23
```

File: package.json [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    "start": "node index.js",
+    "pretest": "eslint . --quiet",
+    "test": "nyc --report-dir ./artifacts/coverage --reporter=lcov mocha --reporter mocha-multi-reporters --reporter-options configFile=./mocha.config.json --recursive --timeout 4000 --retries 1 --exit --allow-uncaught true --color true",
+    "semantic-release": "semantic-release"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/screwdriver-cd/artifacts-unzip-worker.git"
+  },
+  "author": "",
```
Comment (tk3fftk) [inline package.json]: author is empty 👀

File: screwdriver.yaml [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+  # Deploy to beta environment and run tests
+  beta:
+    image: node:16
+    requires: [docker-publish]
+    steps:
+      - setup-ci: git clone https://github.com/screwdriver-cd/toolbox.git ci
+      - wait-docker: DOCKER_TAG=`meta get docker_tag` ./ci/docker-wait.sh
+      - deploy-k8s: K8S_TAG=`meta get docker_tag` ./ci/k8s-deploy.sh
+      - test: npm install && npm test
```
Comment (tk3fftk) [inline screwdriver.yaml]: Do we need unit test after deploying?

File: test/lib/jobs.test.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+            try {
+                await jobs.unzip.perform(unzipConfig);
+                assert.fail('Never reaches here');
+            } catch (err) {
+                assert.equal(err.message, 'failed to get ZIP artifacts from Store');
+            }
+        });
+
+        it('raises an error when it failed to get ZIP artifacts', async () => {
```
Comment (tk3fftk) [inline test/lib/jobs.test.js]:
```suggestion
        it('raises an error when it failed to put ZIP artifacts', async () => {
```

File: config/custom-environment-variables.yaml (hunk 4 above)
Comment (sakka2) [reply config/custom-environment-variables.yaml]: That's right. I fixed to unzip-worker.

File: lib/helper/request-store.js (hunk 5 above)
Comment (sakka2) [reply lib/helper/request-store.js]: I fixed to make const.

File: package.json (hunk 7 above)
Comment (sakka2) [reply package.json]: I fixed.

File: screwdriver.yaml (hunk 8 above)
Comment (sakka2) [reply screwdriver.yaml]: Not needed. I deleted.

**Comment by sakka2** [comment]
> Now, [Retry Plugin](https://github.com/sonic-screwdriver-cd/artifacts-unzip-worker/blob/8dde7c7355c4ca2aed46d0759d6f620f513a0bbd/lib/jobs.js#L52) dose not work. In the description, I added the unzip-worker's log and redis-server's keys status while the job fails. Does anyone have any idea with this problem?

This problem is solved.
[Scheduler](https://github.com/actionhero/node-resque#scheduler) was needed to execute the  reEnqued job (Delayed Job Queue）.
[The scheduler is created by queue-service](https://github.com/screwdriver-cd/queue-service/blob/master/plugins/queue/scheduler.js#L489), so the retry process was executed properly by running the queue-service together.

**Review by tkyi — APPROVED** [review]

**Comment by yuichi10** [comment]
I would like someone with authority to create a pipeline for this repository 🙏

**Review by s-yoshika — APPROVED** [review]
LGTM (I have no permissions to create pipeline)

File: screwdriver.yaml [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      SD_SONAR_OPTS: "-Dsonar.sources=lib,plugins,helpers -Dsonar.tests=test -Dsonar.javascript.lcov.reportPaths=artifacts/coverage/lcov.info -Dsonar.testExecutionReportPaths=artifacts/report/test.xml"
+    requires: [~pr, ~commit]
+    steps:
+      - install: npm install
+      - test: npm test
+      - coverage: ./ci/coverage.sh
+      - cp-coverage-to-artifacts: cp -r artifacts/coverage $SD_ARTIFACTS_DIR
+
+  # Publish the package to GitHub and build Docker image
+  publish:
```
Comment (tkyi) [inline screwdriver.yaml:16]: missing `image`

Comment (yuichi10) [reply screwdriver.yaml:16]: I fixed.

File: screwdriver.yaml [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+      # Publishing to NPM
+      - NPM_TOKEN
+      # Pushing tags to Git
+      - GIT_KEY
+
+  docker-publish:
+    requires: publish
+    template: sd/dind@latest
+
+  # Deploy to beta environment and run tests
```
Comment (yuichi10) [inline screwdriver.yaml:33]: I comment out deploy jobs because it does not work until the all sd-zip-artifacts function has come. It just takes resources.

**Comment by yuichi10** [comment]
@tkyi 
Thank you for creating the pipeline!

All commented items have been corrected.

**Review by tkyi — APPROVED** [review]

**Pull request merged**